Guard

WanchengWildMan opened the pull request:
guard附带一个简单的filter和pipe
每个接口的角色需要确定
日志的格式及数量

WanchengWildMan committed:
http签名1.0

WanchengWildMan committed:
http签名未RFC编码+keyservice1.0

WanchengWildMan committed:
http签名1.0+根据ak验证role,nest can't resolve keyservice 未解决

WanchengWildMan committed:
http签名未rfc编码+guard改写+未指定响应中的message

WanchengWildMan committed:
req.queryreq.headers转换为小写及RFC3986编码
删除样版代码
权限认证采用等级数比较方式

WanchengWildMan committed:
key.controller:canel-

WanchengWildMan committed:
解决在其他控制器加入时 can't resolve dependencies:修改keymodule和authmodule以及appmodule的module文件
定义keyLength
log:将时间删除，加入管理员访问log
开始写过滤器/status处理
key.controller:只能添加非root密钥对

WanchengWildMan committed:
大量规范

WanchengWildMan committed:
修复get和generate中的错误
get没存找到的sk
crypto生成的带标头

WanchengWildMan committed:
删除垃圾文件

WanchengWildMan committed:
重命名roles为decoratos

WanchengWildMan committed:
加入验证管道，验证query参数等

WanchengWildMan committed:
修复guard中check函数返回promise时未解封成boolean

WanchengWildMan committed:
修改日志格式，
删除使用redis事务，返回的值的类型未知

WanchengWildMan committed:
把刚才测试屏蔽掉的需要guard的接加回去

WanchengWildMan committed:
权限删除时逻辑未完全明确

WanchengWildMan committed:
取消使用redis事务

WanchengWildMan committed:
改正generateAddKeyPair的注释

WanchengWildMan committed:
为方法加装控制器

WanchengWildMan committed:
将key.controller中的初始化移到service中,取消未给定角色直接放行,调整部分日志格式,注释中参数,TODO:明确各接口的角色

WanchengWildMan committed:
rename:将del改为remove
pipe校验变量名为count的number

Nugine commented:
PR #19 的合并顺位在 #17 之后。
内容比较多，我可能过两天才有时间看完。

WanchengWildMan committed:
修复lint问题和format问题

WanchengWildMan committed:
改写auth.filter中的长三目运算为if

A0nameless0man commented on POST:
```diff
@@ -0,0 +1,2 @@
+[master 94318d4] key.controller:canel-
+ 4 files changed, 24 insertions(+), 11 deletions(-)
```
这是啥？

A0nameless0man commented on package.json:
```diff
@@ -1,88 +1,89 @@
-        "test:cov": "jest --coverage",
-        "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
-        "test:e2e": "jest --config ./test/jest-e2e.json"
+  "name": "heng-controller",
```
发生了啥？

WanchengWildMan committed:
非root密钥长度改为64，删除无用的FIXME，修正packages里面的格式化风格，删除未知来源的POST文件

A0nameless0man commented on src/auth/auth.guard.ts:
```diff
@@ -0,0 +1,174 @@
+        context: ExecutionContext
+    ): boolean | Promise<boolean> | Observable<boolean> {
+        const rolesRequired: string[] = this.reflector.get(
+            "roles",
```
1. 建议搞个私有或者公开的常量来代替代码中的字符串
2. 加个前缀以防止冲突比较好

A0nameless0man commented on src/auth/auth.guard.ts:
```diff
@@ -0,0 +1,174 @@
+
+        // if (this.whiteUrlList.indexOf(req.url) != -1) return true;
+        //验证http请求头及签名
+        const accessKey = req.headers[PublicHeadersType.accesskey] as string;
```
检查一下类型以避免 `as` ？

A0nameless0man commented on src/auth/dto/roles.dto.ts:
```diff
@@ -0,0 +1 @@
+
```
空文件?

Nugine replied:
<del>空文件建议删除，真要用的话再加。</del>

不知道为什么这条评论不能 resolve，就手动加一条删除线了

Handgrip commented on src/auth/auth.decl.ts:
```diff
@@ -0,0 +1,50 @@
+export type KeyPair = {
+    ak: string | null;
+    sk: string | null;
+    roles?: string[] | null;
```
RoleTypeArr[]

Handgrip commented on src/auth/auth.filter.ts:
```diff
@@ -0,0 +1,32 @@
+        }
+        const msgLog = {
+            statusCode: status, // 系统错误状态
+            timestamp: new Date().toISOString(), // 错误日期
```
https://github.com/ThinkSpiritLab/heng-controller/pull/17#discussion_r584230228

Handgrip commented on src/auth/auth.guard.ts:
```diff
@@ -0,0 +1,174 @@
+        if (!this.checkPermissionValid(rolesRequired, keyPair.roles)) {
+            this.logger.error(
+                `权限不足! accessKey:${accessKey.substring(0, 6)}... ip:${
+                    req.ip
```
忘了NGINX代理之后ip是什么了，可能是127.0.0.1，所以要用header['x-forward-for']或者header['x-real-ip']

Nugine replied:
其实我觉得 IP 不是很重要，因为穿越多层转发时不一定会带上真实 IP，能保证拿到的 IP 只有部署机器上看到的远端 IP。

A0nameless0man replied:
我觉得作为log打一下还是有排查的意义。

Handgrip reviewed (commented):
好像缺文档

WanchengWildMan committed:
添加auth.decl中的注释/文档+修复KeyService中的密钥对生成方法存在/字符+修改req.ip为req.headers["x-forwarded-for"]

WanchengWildMan committed:
增加注释，更换interface dictionary的名称为Dictionary

Nugine commented on src/auth/auth.decl.ts:
```diff
@@ -0,0 +1,50 @@
+export interface dictionary<T = any> {
+    [key: string]: T;
+}
```
有内置 Record 类型可用

Nugine commented on src/auth/auth.decl.ts:
```diff
@@ -0,0 +1,50 @@
+    [key: string]: T;
+}
+export const RoleTypeArr = ["root", "admin", "judger", "user"];
+//buya9o等级比较
```
这行是啥

Nugine commented on src/auth/auth.decl.ts:
```diff
@@ -0,0 +1,50 @@
+//buya9o等级比较
+export const RoleLevel: dictionary = {
+    root: 3,
+    admin: 2,
```
admin 是指外部系统吗？如果是，建议修改命名。
我们的评测系统对外部系统提供服务，外部没有管理权限。

Nugine commented on src/auth/auth.decl.ts:
```diff
@@ -0,0 +1,50 @@
+export const keyPoolsNames = keyPoolsNamesDicTemp;
+export const roleType = roleTypeDicTemp;
+export const toPoolName = keyPoolsNames;
+export const toRoleName = toRoleNameTemp;
```
这里既然使用了多个常量定义，建议修改命名法为 SHOUTY_SNAKE_CASE，例如 KEY_LENGTH。
成员变量一般用小驼峰命名，全局常量用全大写加下划线分隔。

Nugine commented on src/auth/dto/key.dto.ts:
```diff
@@ -0,0 +1,18 @@
+export class KeyPairDto {
+    @IsString()
+    @IsNotEmpty()
+    @Length(keyLength,100, { message: lengthErrorMessage })
```
长度限制 100 是从密钥生成规则导出的吗？

Nugine commented on config/application.toml:
```diff
@@ -29,4 +29,7 @@ rpcTimeout = 3000
+backupRestoreInterval = 5000
+[rootKeyPair]
+rootAccessKey= "w4567t8yhuygft65789j45e6r7t8tgyud45we6575drtfyguft5d46r689gftydr8t90yytd7r9yougiuydtg12"
+rootSecretKey= "5ere56r7t89huytcrd6f789hugyctr678t9y657g8uhinuyt7fgugyftdr5rs64e57r68t79y80uyfd5e7r8yft"
```
部署时如何手动生成 root 密钥？

Nugine commented on src/auth/auth.guard.ts:
```diff
@@ -0,0 +1,174 @@
+        // {body hash}\n
+        const bodyHash = crypto
+            .createHash("sha256")
+            .update(req.body ? JSON.stringify(req.body) : "")
```
JSON.stringify 一个对象时，得到的输出不能保证一致。签名方与校验方的计算结果可能不同。
应该用未经解析的 HTTP body 来计算哈希。

Nugine commented on src/auth/key/key.controller.ts:
```diff
@@ -0,0 +1,96 @@
+
+@Controller("key")
+@UseGuards(RoleSignGuard)
+export class KeyController {
```
@A0nameless0man  我们直接以这个 PR 合并时的密钥接口实现作为协议标准吧

A0nameless0man replied:
好

A0nameless0man replied:
那么在合并之前我需要调整评测机那边的算法确认可以对接。

WanchengWildMan committed:
KeyPairDto 中的长度限制定为[64,64]
常量keyLength改名为KEY_LENGTH_NOT_ROOTOODictionary改为Record<string,string>
admin 改名?
纠正signature的拼写错误
改bodyhash中的JSON.stringfy使用未经解析的body参数添加各个角色的常量
密钥对生成测试接口
文档1.0

WanchengWildMan commented:
昨天github一直上不去，，，然后今天就一起交了，大部分已经改了）

WanchengWildMan commented:
那么admin指的应该是谁的管理权限呢?外部系统没有管理权限，那指的是内部吗?

WanchengWildMan committed:
no-auth注解，test/add接口，key.service中使用DTO

WanchengWildMan commented:
传输密钥时的加密措施是在哪里实现的啊）

WanchengWildMan committed:
增加/add和/getall测试接口，fixme:req.body会报错

Nugine commented:
> 那么admin指的应该是谁的管理权限呢?外部系统没有管理权限，那指的是内部吗?

根据 #11 的描述，只有 root 才能管理整个评测系统。外部系统和评测机都无权管理。

Nugine commented:
> 传输密钥时的加密措施是在哪里实现的啊）

是指生成密钥时的那个 HTTP 响应吗？可以用 root 密钥来加密。

Nugine commented:
#17 现在已经合并，用 git rebase 或 git merge 来解决冲突

WanchengWildMan committed:
lint-fix

WanchengWildMan committed:
改Admin为Root

WanchengWildMan committed:
todo：用root密钥对加密生成的密钥对

WanchengWildMan commented:
key.service.ts
```typescript
    processKey(key: string) {
        return key
            .split("\n")
            .slice(1, 4)
            .join("")
            .replace("/", "A")
            .replace("+", "B")
            .substring(0, KEY_LENGTH_NOT_ROOT);
    }
    async generateKeyPair(roles: string[]): Promise<KeyPairDto> {
        let { publicKey, privateKey } = generateKeyPairSync("rsa", {
            modulusLength: KEY_LENGTH_NOT_ROOT * 16,
            publicKeyEncoding: {
                type: "spki",
                format: "pem"
            },
            privateKeyEncoding: {
                type: "pkcs8",
                format: "pem"
            }
        });
        publicKey = this.processKey(publicKey);
        privateKey = this.processKey(privateKey);
        //TODO 用root密钥对加密
        return {
            ak: publicKey,
            sk: privateKey,
            roles: roles
        };
    }
```
emm如果这里的ak和sk需要对数据加密解密而不仅仅是验证签名的话，，是否需要重构构造密钥的方法呢？
比如返回生成/查询的密钥对的时候需要用root的密钥对对body加密和解密，那么root密钥对就不能是两串随机的字符串了吧？

Nugine commented:
假设有个后台网页用来管理评测系统，那么这个后台网页能知道 root 的 AK 和 SK，评测系统也知道 root 的 AK 和 SK。
双方都知道密钥时，传输信息用某种对称加密即可。密钥生成方式不用改。

WanchengWildMan committed:
管道中的包含验证规则的return_values补上，准备加上入餐验证规则auth.schema

WanchengWildMan committed:
元数据role和no-auth改为大写常量

WanchengWildMan committed:
逐个安装管道，准备改接口为body，准备修复guard中rawBody

WanchengWildMan committed:
TODO:获取rawBody，修改key的接口格式

WanchengWildMan committed:
KeyPairDTO中的roles加@IsArray

WanchengWildMan committed:
更统一的接口格式，基本支持批量操作，TODO：完善文档

WanchengWildMan committed:
文档基本完善

WanchengWildMan committed:
常量命名首字母纠正

Nugine commented on src/auth/auth.decl.ts:
```diff
@@ -1,50 +1,84 @@
+export const ROLES_METADATA = `${MetadataPre}:roles`;
+export const NO_AUTH_METADATA = `${MetadataPre}:no-auth`;
+//Temp变量用于初始化要export的变量
+const keyPoolsNamesArrTemp: string[] = [];
```
写个匿名函数就能去掉这些temp了

Nugine commented on docs/认证模块.md:
```diff
@@ -0,0 +1,470 @@
+
+#### 生成方式
+
+crypto库中的`generateKeyPairSync()`,带有PUBLICKEY/SECETKEY等字样的前缀或后缀，需截取中间部分1到4行，取中间`KEY_LENGTH_NOT_ROOT` 位。
```
这是个 typo 吗？
SECETKEY => SECRETKEY

Nugine commented on docs/认证模块.md:
```diff
@@ -0,0 +1,470 @@
+   	  @IsOptional()
+       @IsArray()
+       @IsSubSetOf(RoleTypeDicExceptRoot, true)
+       roles?: string[];
```
之前一个密钥对只能是一种角色，为什么改成了一对多，这里有什么特殊考虑吗

Nugine reviewed (changes requested):
src/auth/node_modules 不应该包含在提交里，可以新加一个提交删掉。

WanchengWildMan committed:
重命名,删除模块中yarn装不上用npm装上去的node_modules,TODO：rawBody的提取,改回一对一

WanchengWildMan committed:
改回密钥对和角色一对一，优化密钥对生成方式(解决rsa要求的modulusLength较长，取开头时重复度过高，pem格式中间含有保留字符)，增 加测试接口中testAdd和findAllByRoles的默认测试数据

WanchengWildMan committed:
修复取出密钥对长度不对的bug，key.service.processKey,修改一部分文档

WanchengWildMan committed:
完善文档，修复test/add接口中浅拷贝的问题

WanchengWildMan committed:
文档debug，将修改后的root密钥对导出到config/下

WanchengWildMan commented:
emm确认了）

WanchengWildMan committed:
package.json-merge

WanchengWildMan committed:
format

WanchengWildMan committed:
lint
